api/v1alpha1/virtualmachinepublishrequest_types.go

```diff
@@ -0,0 +1,530 @@
+	//
+	// Please note an error will be returned if this field is not
+	// set in a namespace that lacks a default publication target.
+	//
+	// A default publication target is a resource with an API version
+	// equal to spec.target.location.apiVersion, a kind equal to
+	// spec.target.location.kind, and has the label
+	// "imageregistry.vmware.com/default".
+	//
+	// +optional
+	Name string `json:"name,omitempty"`
+
+	// APIVersion is the API version of the referenced object.
+	//
+	// +kubebuilder:default=imageregistry.vmware.com/v1alpha1
+	// +optional
+	APIVersion string `json:"apiVersion,omitempty"`
+
+	// Kind is the kind of referenced object.
+	//
+	// +kubebuilder:default=ContentLibrary
+	// +optional
+	Kind string `json:"kind,omitempty"`
+}
+
+// VirtualMachinePublishRequestTarget is the target of a publication request,
+// typically a ContentLibrary resource.
+type VirtualMachinePublishRequestTarget struct {
+	// Item contains information about the name of the object to which
+	// the VM is published, as well as whether or not to overwrite an
+	// existing object with the same name.
+	//
+	// Please note this value is optional and if omitted, the controller
+	// will use spec.source.name + "-image" as the name of the published
+	// item.
+	//
+	// +optional
+	Item VirtualMachinePublishRequestTargetItem `json:"item,omitempty"`
+
+	// Location contains information about the location to which to publish
+	// the VM.
+	//
+	// +optional
+	Location VirtualMachinePublishRequestTargetLocation `json:"location,omitempty"`
+}
+
+// VirtualMachinePublishRequestSpec defines the desired state of a
+// VirtualMachinePublishRequest.
+//
+// All the fields in this spec are optional. This is especially useful when a
+// DevOps persona wants to publish a VM without doing anything more than
+// applying a VirtualMachinePublishRequest resource that has the same name
+// as said VM in the same namespace as said VM.
+type VirtualMachinePublishRequestSpec struct {
+	// Source is the source of the publication request, ex. a VirtualMachine
+	// resource.
+	//
+	// If this value is omitted then the publication controller checks to
+	// see if there is a resource with the same name as this
+	// VirtualMachinePublishRequest resource, an API version equal to
+	// spec.source.apiVersion, and a kind equal to spec.source.kind. If such
+	// a resource exists, then it is the source of the publication.
+	//
+	// +optional
+	Source VirtualMachinePublishRequestSource `json:"source,omitempty"`
+
+	// Target is the target of the publication request, ex. item
+	// information and a ContentLibrary resource.
+	//
+	// If this value is omitted, the controller uses spec.source.name + "-image"
+	// as the name of the published item. Addtionally, when omitted the
+	// controller attempts to identify the target location by matching a
+	// resource with an API version equal to spec.target.location.apiVersion, a
+	// kind equal to spec.target.location.kind, w/ the label
```

Comment:
Same as above, how could we get spec.target.location.kind/apiVersion if the entire target value is omitted?

Reply:
Same as the above answer but look in `VirtualMachinePublishRequestTarget`